lib/vrl/stdlib/src/hmac.rs

```diff
@@ -0,0 +1,226 @@
+
+fn hmac(value: Value, key: Value, algorithm: &Bytes, encoding: &Bytes) -> Resolved {
+    let value = value.try_bytes()?;
+    let key = key.try_bytes()?;
+
+    let code_bytes = match algorithm.as_ref() {
+        b"SHA1" => hmac!(Sha1, key, value),
+        b"SHA-224" => hmac!(Sha224, key, value),
+        b"SHA-256" => hmac!(Sha256, key, value),
+        b"SHA-384" => hmac!(Sha384, key, value),
+        b"SHA-512" => hmac!(Sha512, key, value),
+        _ => unreachable!("enum invariant")
+    };
+
+    let hash = match encoding.as_ref() {
+        b"hex" => hex::encode(code_bytes),
+        b"base64" => base64::encode(code_bytes),
+        _ => unreachable!("enum invariant")
+    };
+    Ok(hash.into())
+}
+
+fn encode_formats() -> Vec<Value> {
+    vec![
+        value!("hex"),
+        value!("base64")
+    ]
+}
+
+fn algorithms() -> Vec<Value> {
+    vec![
+        value!("SHA1"),
+        value!("SHA-224"),
+        value!("SHA-256"),
+        value!("SHA-384"),
+        value!("SHA-512"),
+    ]
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Hmac;
+
+impl Function for Hmac {
+    fn identifier(&self) -> &'static str {
+        "hmac"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+
+            Parameter {
+                keyword: "key",
+                kind: kind::BYTES,
+                required: true,
+            },
+
+            Parameter {
+                keyword: "algorithm",
+                kind: kind::BYTES,
+                required: false
+            },
+
+            Parameter {
+                keyword: "encoding",
+                kind: kind::BYTES,
+                required: false
+            }
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "default (SHA-256, base64-encoded result)",
+                source: r#"hmac("Hello there", "supersecretkey")"#,
+                result: Ok("kmpc79vrb6SODvg4LwivUnb443+IhR9SSW55KcBPKo8="),
+            },
+
+            Example {
+                title: "SHA-256, hex encoded result",
+                source: r#"hmac("Hello there", "supersecretkey", encoding: "hex")"#,
+                result: Ok("926a5cefdbeb6fa48e0ef8382f08af5276f8e37f88851f52496e7929c04f2a8f")
+            },
+
+            Example {
+                title: "SHA1, base64-encoded result",
+                source: r#"hmac("Hello there", "supersecretkey", algorithm: "SHA1")"#,
+                result: Ok("795HKoopDtOb45EOUroxeHz1OWo=")
+            }
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::TypeState,
+        _ctx: &mut FunctionCompileContext,
+        arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let key = arguments.required("key");
+        let algorithm = arguments

```

Comment:
`algorithm` here can just be kept as `Box<dyn Expression>`. (Since checking the algorithm in `hmac` is necessary anyway, there's no benefit to also trying to check it at compile time here).

Also just a note, but `try_bytes` is not appropriate here. `try_bytes` checks if the _expression_ passed in is a literal `Bytes` expression, but this is called before the value has been resolved / evaluated, so you could have a valid expression that resolves to `Bytes` panic here.

Reply:
I would also suggest keeping it an an `Option` instead of applying the default here. The default can be applied in the `hmac` function above. (Otherwise you would need the default to be a boxed expression, rather than just `None`).